Nadro/2608/sketch mode scene state improvements

nadr0 opened the pull request:
closes https://github.com/KittyCAD/modeling-app/issues/2608

- [recreating the issues](https://github.com/KittyCAD/modeling-app/issues/2608#issuecomment-2341619453)

#### Bugs

- When cancelling the sketch mode from the `Exit Sketch` button or pressing `ESC` while in the sketch mode, we do not exit the sketch mode via `sketch_mode_disable`. This keeps the axis labels around.
- When using the code editor it can batch send the engine command `enable_sketch_mode` but does not send `sketch_mode_disable` when there is no code. This keeps the axis labels around.
- When pressing `ESC` in `sketch -> line tool -> no points` it does not call `Cancel`. It called `change tool`. This was problematic because we could not exit the sketch mode properly. 

#### Expected workfows

##### Workflow 1
`Load into blank KCL file` -> `Click Start Sketch` -> `Click XZ plane` ->  `Click Exit Sketch`
End state: Code should be deleted, axis labels should be hidden.

##### Workflow 2
`Load into blank KCL file` -> `Click Start Sketch` -> `Click XZ plane` ->  `Hit ESC`
End state: Code should be deleted, axis labels should be hidden. 
Should immediately exit the sketch mode, not change your tool. 

##### Workflow 3
- Copy and paste partially complete code into the editor 
```
const sketch001 = startSketchOn('XZ')
  |> startProfileAt([0.07, 0.15], %)
  |> line([0.27, 0.43], %)
```
- Select all the code and delete it
End state: Code should be deleted, axis labels should be hidden.


##### E2E tests
I made an [update](https://github.com/KittyCAD/modeling-app/pull/3866/commits/ea47f10e1ec3e7dece33b89e270b5275b84eb98e) to an E2E test.

This was required because there was a bug that if you clicked the line tool when it was activated and there were no points it wouldn't actually close sketch. The user can just click right into rectangle tool. 

Broken workflow:
Start Sketch Mode -> automatically select line tool -> select line tool with `no points` -> exit line tool -> stay in sketch mode

New workflow:
Start Sketch Mode -> automatically select line tool -> select line tool with `no points` -> exit sketch mode since there are `no points` -> idle

franknoirot commented on src/lib/toolbar.ts:
```diff
@@ -266,10 +266,11 @@ export const toolbarConfig: Record<ToolbarModeName, ToolbarMode> = {
       {
         id: 'sketch-exit',
-        onClick: ({ modelingSend }) =>
+        onClick: ({ modelingSend }) => {
```
If you want a lint rule or prettier config for this (all functions must use curly braces) I wouldn't be opposed, it's not the first time it's tripped one of us up. Last time it was an `if` statement with no braces and an early return that wrapped to the next line.

nadr0 replied:
I can revert this. I think I had some debug code and deleted it but forgot to remove the syntax. I am fine with our current pattern.

franknoirot commented on src/lang/KclSingleton.ts:
```diff
@@ -293,6 +293,12 @@ export class KclManager {
     }
+
+    // Exit sketch mode if the AST is empty
+    if (this._isAstEmpty(ast)) {
+      await this.disableSketchMode()
+    }
```
This `executeAst` function is probably running when you first enter your first sketch, with an empty AST, and throwing a wrench in things from the jump in certain Playwright tests. I still agree with the approach, but does this check maybe need to live in a different place?

nadr0 replied:
Thanks for the tip, I'll take a look at this.

nadr0 commented:
Updated one more E2E test to account for the new `no points` workflow, [commit](https://github.com/KittyCAD/modeling-app/pull/3866/commits/e7380801fc918a3dbc18a980a96bf481600807e0)

To test the same workflow of `Deselecting line tool should mean nothing happens on click` we need to place at least one point during sketch mode. If we do not place one point the sketch mode will be exited and the code will be cleared. I've added a base point then the tool can be turned on and off to see that no content changes. 

![Screenshot from 2024-09-19 15-17-51](https://github.com/user-attachments/assets/8ead1662-5be3-4087-acd4-770bb586d8e3)

Irev-Dev commented:
Dope.
